src/backend/distributed/utils/metadata_cache.c

```diff
@@ -276,99 +402,172 @@ LookupDistTableCacheEntry(Oid relationId)
 {
 	DistTableCacheEntry *cacheEntry = NULL;
 	bool foundInCache = false;
-	HeapTuple distPartitionTuple = NULL;
-	char *partitionKeyString = NULL;
-	char partitionMethod = 0;
-	uint32 colocationId = INVALID_COLOCATION_ID;
-	char replicationModel = 0;
-	List *distShardTupleList = NIL;
-	int shardIntervalArrayLength = 0;
-	ShardInterval **shardIntervalArray = NULL;
-	ShardInterval **sortedShardIntervalArray = NULL;
-	FmgrInfo *shardIntervalCompareFunction = NULL;
-	FmgrInfo *hashFunction = NULL;
-	bool hasUninitializedShardInterval = false;
-	bool hasUniformHashDistribution = false;
 	void *hashKey = (void *) &relationId;
-	Relation pgDistPartition = NULL;
 
 	if (DistTableCacheHash == NULL)
 	{
 		InitializeDistTableCache();
 	}
 
-	cacheEntry = hash_search(DistTableCacheHash, hashKey, HASH_FIND, &foundInCache);
+	cacheEntry = hash_search(DistTableCacheHash, hashKey, HASH_ENTER, &foundInCache);
 
 	/* return valid matches */
-	if ((cacheEntry != NULL) && (cacheEntry->isValid))
+	if (foundInCache)
 	{
-		return cacheEntry;
-	}
+		if (cacheEntry->isValid)
+		{
+			return cacheEntry;
+		}
 
-	/* free the content of old, invalid, entries */
-	if (cacheEntry != NULL)
-	{
+		/* free the content of old, invalid, entries */
 		ResetDistTableCacheEntry(cacheEntry);
 	}
 
+	/* zero out entry, but not the key part */
+	memset(((char *) cacheEntry) + sizeof(Oid), 0,
+		   sizeof(DistTableCacheEntry) - sizeof(Oid));
```

Comment:
Just for cleanliness, could we reset the Oid after setting the whole struct to 0?